### PR TITLE
gnutls bugfix: Fix error handling in gtlsRecordRecv

### PR DESCRIPTION
**Pull request by alorbach**
There was a rare possibility that the E_AGAIN/E_INTERRUPT handling
could cause an infinite loop (100% CPU Usage), for example when a TLS
handshake is interrupted at a certain stage.

- After gnutls_record_recv is called, and E_AGAIN/E_INTERRUPT error
  occurs, we need to do additional read/write direction handling
  with gnutls_record_get_direction.
- After the second call of gnutls_record_recv (Expand buffer)
  we needed to also check the eror codes for E_AGAIN/E_INTERRUPT
  to do propper errorhandling.

see also: https://github.com/rsyslog/rsyslog/issues/4818

### CODE REVIEW

File: runtime/nsd_gtls.c [hunk 1]
```diff
@@ -576,14 +577,31 @@ gtlsRecordRecv(nsd_gtls_t *pThis)
...
 		}
 	} else if(lenRcvd == GNUTLS_E_AGAIN || lenRcvd == GNUTLS_E_INTERRUPTED) {
-sslerr:
-		pThis->rtryCall = gtlsRtry_recv;
-		dbgprintf("GnuTLS receive requires a retry (this most probably is OK and no error condition)\n");
-		ABORT_FINALIZE(RS_RET_RETRY);
+sslerragain:
+		/* Check if the underlaying file descriptor needs to read or write data!*/
+		wantsWriteData = gnutls_record_get_direction(pThis->sess);
+		if (wantsWriteData == 0) {
```
Comment (lonicerae) [inline runtime/nsd_gtls.c]: In our case, the problem is the thread keeps doing 'recvfrom()' endlessly. Could we abort the connection after some retries or some time?

**Review by lonicerae — CHANGES_REQUESTED** [review]

File: runtime/nsd_gtls.c (hunk 1 above)
Comment (alorbach) [reply runtime/nsd_gtls.c]: @rgerhards what do you think? In order to maintain existing 

> In our case, the problem is the thread keeps doing 'recvfrom()' endlessly. Could we abort the connection after some retries or some time?

Could you post a gdb stack trace?
Or even better could you enable debug output, if we could catch that error with debug log enabled that would be awesome.

These modules would be sufficient for debug logging: 
```
global( 
       debug.whitelist="on"
       debug.files=["nsd_gtls.c", "nsd_ptcp.c", "tcpsrv.c", "nsdsel_ossl.c", "nsdsel_gtls.c", "nsdpoll_ptcp.c", "dnscache.c"]
)
```

**Comment by opoplawski** [comment]
I have debug output, but it's lot of data and I don't want to post it publicly.

**Comment by opoplawski** [comment]
I'm afraid the debug output is 1.5GB compressed and too large for that form to upload.

**Comment by alorbach** [comment]
> I'm afraid the debug output is 1.5GB compressed and too large for that form to upload.
 I have extended the form to support up to 10GB files. Can you retry?

**Comment by alorbach** [comment]
> I've uploaded the segment. Hope it helps. Sorry for the delay - vacation and some technical issues.

thanks I got it 👍 , I will review the logfile ASAP.

**Comment by alorbach** [comment]
> I've uploaded the segment. Hope it helps. Sorry for the delay - vacation and some technical issues.
I reviewed the debug logfile and this really looks odd. Something is not handeled right at some point, properly with a huge amount of sessions. 

I have added some more debug output and added output of the nsd pointer at more debug outputs so we can track which nsd is doing what. For some reason it never reaches the point of "nsdsel_gtls: reached end".  

@opoplawski  Can you try to run with the amended debug output changes from here? It will help me better tracking the nsd objects inside the debug logfile.

**Comment by opoplawski** [comment]
I've started running it, but it's generating about 1GB of output each minute - so not sure how long I can run it.

**Comment by opoplawski** [comment]
New file uploaded

**Comment by alorbach** [comment]
@opoplawski Thanks a lot we are coming closer ;)

> tcpsrv.c: tcpsrv: ready to process 1 event entries
> imtcp.c        : tcpsrv.c: tcpsrv: processing item 15, pUsr 0x7f133c1bbb60, bAbortConn
> imtcp.c        : tcpsrv.c: netstream 0x7f133c1bba30 with new data
> imtcp.c        : nsd_gtls.c: gtlsRecordRecv: start
> imtcp.c        : nsd_gtls.c: gtlsRecordRecv: gnutls_record_recv received 456 bytes
> imtcp.c        : nsd_gtls.c: gtlsRecordRecv return. nsd 0x7f133c19ab60, iRet 0, lenRcvd 456, lenRcvBuf 456, ptrRcvBuf 0
> imtcp.c        : nsd_gtls.c: gtlsRcv return. nsd 0x7f133c19ab60, iRet 0, lenRcvBuf -1, ptrRcvBuf 456

The same tcp session is causing the loop over and over, however I can't see why. It looks like rsyslog is getting hammered with the same amount of data on that TCP Session over and over again. 

I have added more debug output, it will now reveal the tcp session client, causing the loop. Its ok to have debug log run for 1 minute, I don't need more data. 

Thanks a lot for helping.

**Comment by opoplawski** [comment]
Uploaded new log.  This is just the final minute before the server was shutdown but probably doesn't capture the initial start of the issue.

**Comment by opoplawski** [comment]
@alorbach Is this ready for testing?

**Comment by alorbach** [comment]
@opoplawski : Yes I added a potential fix for the problem. However I cannot test it, if you could run it at your end it would be great. When you debug you may see "doRetry: gtlsRecordRecv returned RETRY, but there is no pending" from time to time, when this happens it could have caused the 100% loop before.

**Comment by alorbach** [comment]
@opoplawski Did you have a chance to test the PR again?

**Comment by opoplawski** [comment]
Yes - it's been running for nearly 2 days and I haven't seen the problem again - which is the longest I've ever seen it go.  So I think this has likely fixed the issue for me.

**Comment by alorbach** [comment]
@opoplawski Thanks a lot for the feedback 👍 ! If it hasn't happened again, it most likely is fixed now. 
Can you check if any rsyslog tls errors werre logged during meantime?

**Comment by alorbach** [comment]
So it appears that this PR should close the following issues: 
https://github.com/rsyslog/rsyslog/issues/4834
https://github.com/rsyslog/rsyslog/issues/4818
https://github.com/rsyslog/rsyslog/issues/4638

**Pull request merged**